lib/CompositorController.js

```diff
@@ -0,0 +1,401 @@
+      callback();
+    }
+  }
+
+  /**
+   * Posts a task to issue a BeginFrame while waiting for the
+   * mainFrameReadyForScreenshots event. The taks may be cancelled by the event.
+   */
+  _postWaitForCompositorReadyBeginFrameTask() {
+    // We may receive the mainFrameReadyForScreenshots event before this task
+    // is run. In that case, we cancel it in _mainFrameReadyForScreenshots to
+    // avoid another unnecessary BeginFrame.
+    this._waitForCompositorReadyBeginFrameTask = setTimeout(
+        this._issueWaitForCompositorReadyBeginFrame.bind(this),
+        this.waitForCompositorReadyBeginFrameDelay);
+  }
+
+  _issueWaitForCompositorReadyBeginFrame() {
+    // No need for PostBeginFrame, since _waitForCompositorReadyBeginFrameTask
+    // has already been posted.
+    if (this._waitForCompositorReadyBeginFrameTask) {
+      clearTimeout(this._waitForCompositorReadyBeginFrameTask);
+      this._waitForCompositorReadyBeginFrameTask = null;
+    }
+
+    this._beginFrame(this._waitForCompositorReadyBeginFrameComplete.bind(this));
+  }
+
+  _waitForCompositorReadyBeginFrameComplete() {
+    if (this._mainFrameReady) {
+      const compositorReadyCallback = this._compositorReadyCallback;
+      this._compositorReadyCallback = null;
+      compositorReadyCallback();
+      return;
+    }
+
+    // Continue posting more BeginFrames with a delay until the main frame
+    // becomes ready. If needs_begin_frames_ is false, it will eventually turn
+    // true again once the renderer's compositor has started up.
+    if (this._needsBeginFrames)
+      this._postWaitForCompositorReadyBeginFrameTask();
+  }
+
+
+  /**
+   * Posts a BeginFrame while waiting for a main frame content update.
+   */
+  _postWaitForMainFrameContentUpdateBeginFrame() {
+    this._postBeginFrame(this._waitForMainFrameContentUpdateBeginFrameComplete.bind(this));
+  }
+
+  _waitForMainFrameContentUpdateBeginFrameComplete(beginFrameResult) {
+    if (!beginFrameResult)
+      return;
+
+    if (beginFrameResult.mainFrameContentUpdated) {
+      const callback = this._mainFrameContentUpdatedCallback;
+      this._mainFrameContentUpdatedCallback = null;
+      callback();
+      return;
+    }
+
+    // Continue posting BeginFrames until we see a main frame update.
+    if (this._needsBeginFrames)
+      this._postWaitForMainFrameContentUpdateBeginFrame();
+  }
+
+  _captureScreenshotBeginFrameComplete(beginFrameResult) {
+    if (beginFrameResult && beginFrameResult.screenshotData) {
+      const callback = this._screenshotCapturedCallback;
+      this._screenshotCapturedCallback = null;
+      callback(beginFrameResult.screenshotData);
+    } else {
+      // TODO(alexclarke): this should really be promise rejection.
+      const callback = this._screenshotCapturedCallback;
+      this._screenshotCapturedCallback = null;
+      callback();
+    }
+  }
+}
+
+module.exports = CompositorController;
+helper.tracePublicAPI(CompositorController);
```

Comment:
This line is not needed since `CompositorController` is not exposed in the API

Reply:
Done